dask_expr/expr.py

```diff
@@ -616,6 +618,61 @@ def visualize(self, filename="dask-expr.svg", format=None, **kwargs):
         graphviz_to_file(g, filename, format)
         return g
 
+    def _partitioning(self, columns: list) -> dict:
+        """Known partitioning information
+
+        Return known-partitioning information for the specified
+        list of columns. This information should be formatted
+        as a dict containing "columns" and "how" keys, where
+        the `"columns"` value should be a tuple of column names,
+        and the `"how"` value should be a tuple that uniquely
+        identifies the partitioning.
+
+        If no partitioning information is known, an empty
+        dictionary will be returned.
+
+        Examples of `"how"`:
+
+          - Sorted data: `("increasing", <divisions>)`
+          - Reverse-sorted data: `("decreasing", <divisions>)`
+          - Shuffled data: `("hash", <npartitions>)`
+
+        Note that un-named index columns must be specified as
+        `"__index__"` (`None` is not supported).
+
+        Return
+        ------
+        partitioning: dict
+        """
```

Comment:
I'd welcome a conversation about this.  

My initial thought is that it made sense to store some baseline information like ...

1.  row counts of each partition
2. min/max values of each column in each partition

These are similar to what comes out of parquet.  Then, when we wanted to ask something, we would consult that raw data.  

This feels like we're now storing derivative values off of that data.  This makes me slightly nervous because it opens the door to tracking lots of state.  I would be more comfortable if we were to track the underlying state (counts, mins, maxes) and then decided to compute quantities like these on the fly.  That feels more tightly scoped to me.  

Thoughts?

Reply:
This PR does a few different things, and some of those "things" I am much more confident in than others.

1. We add an optional `_lengths` attribute the `Expr` to store "known" partition lengths.
2. We add an optional `_partitioning` method to `Expr` so that an expression can check if the underlying collection is partitioned by a specific set of columns (even if that column does not include an index with known divisions).
3. We add logic to some `Expr` classes (mostly `ReadParquet`) to "lazily" collect the necessary statistics when `_lengths` or `_partitioning` information is requested.

The primary reason this PR is still marked as "draft" is that the current iteration will **always** attempt to go back and collect statistics in `ReadParquet` when `_lengths` or `_partitioning` are called (and the necessary statistics are missing). While it will always make sense to collect partition-length statistics in support of something like `len(df)`, it may not always be the best idea to collect statistics.  In fact, I'm already a bit uncomfortable with the fact that column-projection and predicate-pushdown optimizations currently require us to repeat the initial dataset processing, which can be slow on some systems (this is something I'd like to address separately).

Note that I also think the specific API can be improved, but the "eagerness" of the lazy-metadata collection feels like the most challenging short-term blocker.

What you seem to be uncomfortable with is the fact that we are not adding something like `Expr._mins` and `Expr._maxes`, but are instead exposing a method to provide more general (derivative) information about how the collection is partitioned.  I'm very open to other approaches. My current proposal here was just the natural result of attempting to store mins/maxes, and finding that my personal attempt at doing so was not particularly clean or useful.  In most cases, the original `ReadParquet` expression will not collect useful min/max statistics.  When the expression **does** collect min/max statistics, the only reason we care about them is to tell us how/if the collection is partitioned.  For this reason, I found it most natural to allow specific classes (like `ReadParquet` and `Shuffle`) to worry about what kinds of statistics they want to collect/track (if any).

I'll think a bit more about this.

Reply:
> I'm already a bit uncomfortable with the fact that column-projection and predicate-pushdown optimizations currently require us to repeat the initial dataset processing, which can be slow on some systems (this is something I'd like to address separately).

I noticed this recently.  I wonder if the parquet code could benefit from a module-level lru-cache

Reply:
> What you seem to be uncomfortable with is the fact that we are not adding something like Expr._mins and Expr._maxes, but are instead exposing a method to provide more general (derivative) information about how the collection is partitioned

Yeah, I'm comparing this to database world where you have a reference table which is the single point of truth (SPOT) and then views on that table.  This feels like we're storing the views as concrete tables.  Bad things tend to result from that behavior.

As an example.  I could imagine future applications aside from sortedness.  We've mentioned a couple of these including filtering / partition pruning and optimizations that are based on the values.  I think that storing the underlying data is more future-proof.  

I probably wouldn't have separate protocols for `_maxes` and `_mins` but maybe a protocol that includes `_min_maxes` or all column-based statistics (if they're likely to be consistent across all systems that provide this information (it might make sense to look at what Snowflake, Parquet, and Delta all provide, for example)).